File: Singly Linked List/singly_Linked_List.js
class Node {
  constructor(val) {
    this.val = val;
    this.next = null;
  }
}

class SinglyLinkedList {
  constructor() {
    this.head = null;
    this.tail = null;
    this.length = 0;
  }
  /*
  This Takes a val to be added to list
  Create a new Node Using the value passed by using the Node class.
  check for head value if it's empty add the new val reference to head 
  and the same to tail.
  if head is not empty.
  change the tail to next and assign newNode to it
  increment length every time a val is added
  return the list at end.
  */
  push(val) {
    let newNode = new Node(val); // A Node containd val and next ref
    if (!this.head) {
      this.head = newNode;
      this.tail = newNode;
    } else {
      this.tail.next = newNode;
      this.tail = newNode;
    }
    this.length++;
    return this;
  }
  /*
  check for empty list 
  set the current to present head and new tail to same head
  check for current.next 
  set new tail to current(this.head) - this is always one node behind current
  set current to the current next
  at end when there are no next
  assign new tail as this.tail
  and newTail next as null
  decrement length
  and return the current (removed val)
  */
  pop() {
    if (!this.head) return undefined;
    let current = this.head;
    let newTail = current;
    while (current.next) {
      newTail = current;
      current = current.next;
    }
    this.tail = newTail;
    newTail.next = null;
    this.length--;
    if (this.length === 0) {
      this.head = null;
      this.tail = null;
    }
    return current;
  }
  /*
  shift means removing the first element from list
  check if the list is empty if empty return undefined
  store the head to new variable
  and make the head next as new head
  decrement length
  */
  shift() {
    if (!this.head) return undefined;
    let current = this.head;
    this.head = current.next;
    this.length--;
    return this;
  }
  /*
  add element at starting of the list
  if the list is empty call push method which sets head and tail as newNode
  store the new val as new Node variable
  and make current head as new Node next
  increment length
  */
  unshift(value) {
    if (!this.head) {
      this.push(value);
    } else {
      let newNode = new Node(value);
      newNode.next = this.head;
      this.head = newNode;
    }
    this.length++;
    return this;
  }
  /*
  if index < 0 or index > this,length return null;
  if index is 0 return head.val
  create variable called counter and assign 0
  set this.head to new variable current
  start looping for current.next 
  assign current = current.next 
  increment counter
  while counter value is equal to index val break
  return the current.val
  */
  getValByIndex(index) {
    if (index < 0 || index >= this.length) return null;
    if (index === 0) return this.head;
    let count = 0;
    let current = this.head;
    while (current.next) {
      current = current.next;
      count++;
      if (count === index) break;
    }
    return current;
  }
  /*
  change value by index
  check for node existance using getValByIndex
  if exist
  change it's value with new value
  return true if changed or false if not changed
  */
  setValByIndex(val, index) {
    let requestedindexCurrent = this.getValByIndex(index);
    if (requestedindexCurrent) {
      requestedindexCurrent.val = val;
      return true;
    }
    return false;
  }
  insertNewAt(val, index) {
    if (index < 0 || index > this.length) return undefined;
    if (index === this.length) return !!this.push(val);
    if (index === 0) return !!this.unshift(val);
    let newNode = new Node(val);

    let prev = this.getValByIndex(index - 1);
    let temp = prev.next;
    prev.next = newNode;
    newNode.next = temp;
    this.length++;
    return true;
  }
  removeByindex(index) {
    if (index < 0 || index >= this.length) return undefined;
    if (index === 0) return !!this.shift();
    if (index === this.length - 1) return !!this.pop;
    let prevNode = this.getValByIndex(index - 1);
    let removed = prevNode.next;
    prevNode.next = removed.next;
    this.length--;
    return true;
  }
  print() {
    let arr = [];
    let current = this.head;
    while (current.next) {
      arr.push(current);
      current = current.next;
    }
    return arr;
  }
  reverse() {
    // return this.print().reverse();
    let node = this.head;
    this.head = this.tail;
    this.tail = node;
    let next;
    let prev = null;
    for (let i = 0; i < this.length; i++) {
      next = node.next;
      node.next = prev;
      prev = node;
      node = next;
    }
    return this;
  }
}

const list = new SinglyLinkedList();
console.log(list.push(0));
console.log(list.push(1));
console.log(list.push(2));
console.log(list.push(3));

console.log(list.pop());
console.log(list.shift());
console.log(list.push(4));
console.log(list.unshift(0));

console.log(list.unshift(5));
console.log(list.getValByIndex(0));
console.log(list.setValByIndex(4, 1));
console.log(list.getValByIndex(1));
console.log(list.insertNewAt(6, 4));
console.log(list.getValByIndex(4));
console.log(list.removeByindex(4));
console.log(list.getValByIndex(4));

console.log("--------------------------");

console.log(list.reverse());

File: Linked List/Singly Linked List/singly_Linked_List.py
class Node:
    def __init__(self, val):
        self.val = val
        self.next = None


class SinglyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.length = 0

    def push(self, val):
        newNode = Node(val)
        if self.head == None:
            self.head = newNode
            self.tail = newNode
        else:
            self.tail.next = newNode
            self.tail = newNode
        self.length += 1
        return self

    def indexOf(self, val):
        if self.head == None:
            return 'No Item Found in List'
        else:
            count = 0
            current = self.head
            if val == current.val:
                return f'Found at index {count}'
            else:
                while current.next:
                    current = current.next
                    count += 1
                    if(current.val == val):
                        return f'Found at index {count}'

                return 'Value is Not Found in the List'

    def pop(self):
        if self.head == Node:
            return "No Item Found To Delete"
        else:
            current = self.head
            while current.next:
                newTail = current
                current = current.next
            current.val = None
            self.tail = newTail
        self.length -= 1
        return self

    def show(self):
        if self.head == None:
            print("List is Empty")
        else:
            current = self.head
            print(current.val)
            while(current.next):
                current = current.next
                if(current.val == None):
                    break
                print(current.val)

    def insertAt(self, val, index):
        if self.head == None:
            self.push(val)
            return self
            self.length += 1
        else:
            count = 0
            current = self.head
            if index == 0:
                current.val = val
                return self
            else:
                while current.next:
                    current = current.next
                    count += 1
                    if(index == count):
                        current.val = val
                        return self
                print('Out Of Index')


s = SinglyLinkedList()
s.push(1)
s.push(2)
s.push(3)
s.pop()
s.push(3)
s.push(4)
s.push(5)
s.push(6)
# s.show()
print(s.indexOf(6))
s.insertAt('One', 0)

s.insertAt('Six', 5)
s.insertAt('Three', 2)
s.show()
s.insertAt('seven', 6)
